File: rclpy/rclpy/subscription.py
# Copyright 2016 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from enum import Enum
import inspect
from types import TracebackType
from typing import Awaitable
from typing import Callable
from typing import Generic
from typing import Literal
from typing import Optional
from typing import overload
from typing import Type
from typing import TYPE_CHECKING
from typing import TypedDict
from typing import TypeVar
from typing import Union


from rclpy.callback_groups import CallbackGroup
from rclpy.event_handler import SubscriptionEventCallbacks
from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
from rclpy.qos import QoSProfile
from rclpy.subscription_content_filter_options import ContentFilterOptions
from rclpy.type_support import MsgT
from typing_extensions import Self
from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from typing_extensions import TypeIs


class PublisherGID(TypedDict):
    implementation_identifier: str
    data: bytes


class MessageInfo(TypedDict):
    source_timestamp: int
    received_timestamp: int
    publication_sequence_number: Optional[int]
    reception_sequence_number: Optional[int]
    publisher_gid: Optional[PublisherGID]


# Re-export exception defined in _rclpy C extension.
RCLError = _rclpy.RCLError

# Left to support Legacy TypeVars.
MsgType = TypeVar('MsgType')

# Can be redone with TypeVar(default=MsgT) when either typing-extensions4.11.0+ or python3.13+
T = TypeVar('T')

GenericSubscriptionCallback: TypeAlias = Union[Callable[[T], None],
                                               Callable[[T, MessageInfo], None]]
AsyncGenericSubscriptionCallback: TypeAlias = Union[Callable[[T], Awaitable[None]],
                                                    Callable[[T, MessageInfo], Awaitable[None]]]
GenericSubscriptionCallbackUnion: TypeAlias = Union[GenericSubscriptionCallback[T],
                                                    AsyncGenericSubscriptionCallback[T]]
SubscriptionCallbackUnion: TypeAlias = Union[GenericSubscriptionCallbackUnion[MsgT],
                                             GenericSubscriptionCallbackUnion[bytes]]

SubscriptionCallback: TypeAlias = Callable[[MsgT | bytes], None] | \
                                  Callable[[MsgT | bytes], Awaitable[None]]
MessageAndInfo: TypeAlias = tuple[MsgT | bytes, MessageInfo]
SubscriptionCallbackWithMessageInfo: TypeAlias = Callable[[*MessageAndInfo[MsgT]], None] | \
                                                 Callable[[*MessageAndInfo[MsgT]], Awaitable[None]]


class BaseSubscription(Generic[MsgT]):

    class CallbackType(Enum):
        MessageOnly = 0
        WithMessageInfo = 1

    @overload
    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: GenericSubscriptionCallbackUnion[bytes],
         qos_profile: QoSProfile,
         raw: Literal[True],
         *,
         on_destroy: Optional[Callable[[Self], None]] = None,
    ) -> None: ...

    @overload
    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: GenericSubscriptionCallbackUnion[MsgT],
         qos_profile: QoSProfile,
         raw: Literal[False],
         *,
         on_destroy: Optional[Callable[[Self], None]] = None,
    ) -> None: ...

    @overload
    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: SubscriptionCallbackUnion[MsgT],
         qos_profile: QoSProfile,
         raw: bool,
         *,
         on_destroy: Optional[Callable[[Self], None]] = None,
    ) -> None: ...

    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: SubscriptionCallbackUnion[MsgT],
         qos_profile: QoSProfile,
         raw: bool,
         *,
         on_destroy: Optional[Callable[[Self], None]] = None,
    ) -> None:
        """
        Create a container for a ROS subscription.

        .. warning:: Users should not create a subscription with this constructor, instead they
           should call :meth:`.Node.create_subscription` or :meth:`.AsyncNode.create_subscription`.
        """
        self.__subscription = subscription_impl
        self.msg_type = msg_type
        self.topic = topic
        self.callback = callback
        self.qos_profile = qos_profile
        self.raw = raw
        self._on_destroy = on_destroy
        self._destroyed = False

    def get_publisher_count(self) -> int:
        """Get the number of publishers that this subscription has."""
        with self.handle:
            return self.__subscription.get_publisher_count()

    @property
    def handle(self) -> '_rclpy.Subscription[MsgT]':
        return self.__subscription

    def destroy(self: Self) -> None:
        """Destroy the subscription, notifying the owning node and releasing the handle."""
        if self._destroyed:
            return
        self._destroyed = True
        if self._on_destroy is not None:
            self._on_destroy(self)
            self._on_destroy = None
        self._destroy()

    def _destroy(self) -> None:
        self.handle.destroy_when_not_in_use()

    @property
    def topic_name(self) -> str:
        with self.handle:
            return self.__subscription.get_topic_name()

    @property
    def callback(self) -> SubscriptionCallbackUnion[MsgT]:
        return self._callback

    @callback.setter
    def callback(self, value: SubscriptionCallbackUnion[MsgT]) -> None:
        self._set_callback_type(value)
        self._callback = value

    @staticmethod
    def _detect_wants_info(callback: SubscriptionCallbackUnion[MsgT]
                           ) -> TypeIs[SubscriptionCallbackWithMessageInfo[MsgT]]:
        try:
            inspect.signature(callback).bind(object())
            return False
        except TypeError:
            pass
        try:
            inspect.signature(callback).bind(object(), object())
            return True
        except TypeError:
            pass
        raise RuntimeError(
            'Subscription callback should be either be callable with one argument'
            '(to get only the message) or two (to get message and message info)')

    def _set_callback_type(self, callback: SubscriptionCallbackUnion[MsgT]) -> None:
        if BaseSubscription._detect_wants_info(callback):
            self._callback_type = BaseSubscription.CallbackType.WithMessageInfo
        else:
            self._callback_type = BaseSubscription.CallbackType.MessageOnly

    @property
    def logger_name(self) -> str:
        """Get the name of the logger associated with the node of the subscription."""
        with self.handle:
            return self.__subscription.get_logger_name()

    @property
    def is_cft_supported(self) -> bool:
        """Check if content filtering is supported for this subscription."""
        with self.handle:
            return self.__subscription.is_cft_supported()

    @property
    def is_cft_enabled(self) -> bool:
        """Check if content filtering is enabled for the subscription."""
        with self.handle:
            return self.__subscription.is_cft_enabled()

    def set_content_filter(self, filter_expression: str, expression_parameters: list[str]) -> None:
        """
        Set the filter expression and expression parameters for the subscription.

        :param filter_expression: The filter expression to set.
        :param expression_parameters: The expression parameters to set.
        :raises: RCLError if internal error occurred when calling the rcl function.
        """
        with self.handle:
            self.__subscription.set_content_filter(filter_expression, expression_parameters)

    def get_content_filter(self) -> ContentFilterOptions:
        """
        Get the filter expression and expression parameters for the subscription.

        :return: ContentFilterOptions object containing the filter expression and expression
            parameters.
        :raises: RCLError if internal error occurred when calling the rcl function.
        """
        with self.handle:
            return self.__subscription.get_content_filter()

    def __enter__(self) -> Self:
        return self

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_val: Optional[BaseException],
        exc_tb: Optional[TracebackType],
    ) -> None:
        self.destroy()


class Subscription(BaseSubscription[MsgT], Generic[MsgT]):

    @overload
    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: GenericSubscriptionCallbackUnion[bytes],
         qos_profile: QoSProfile,
         raw: Literal[True],
         *,
         on_destroy: Optional[Callable[['Subscription[MsgT]'], None]] = None,
         callback_group: CallbackGroup,
         event_callbacks: SubscriptionEventCallbacks,
    ) -> None: ...

    @overload
    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: GenericSubscriptionCallbackUnion[MsgT],
         qos_profile: QoSProfile,
         raw: Literal[False],
         *,
         on_destroy: Optional[Callable[['Subscription[MsgT]'], None]] = None,
         callback_group: CallbackGroup,
         event_callbacks: SubscriptionEventCallbacks,
    ) -> None: ...

    @overload
    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: SubscriptionCallbackUnion[MsgT],
         qos_profile: QoSProfile,
         raw: bool,
         *,
         on_destroy: Optional[Callable[['Subscription[MsgT]'], None]] = None,
         callback_group: CallbackGroup,
         event_callbacks: SubscriptionEventCallbacks,
    ) -> None: ...

    def __init__(
         self,
         subscription_impl: '_rclpy.Subscription[MsgT]',
         msg_type: Type[MsgT],
         topic: str,
         callback: SubscriptionCallbackUnion[MsgT],
         qos_profile: QoSProfile,
         raw: bool,
         *,
         on_destroy: Optional[Callable[['Subscription[MsgT]'], None]] = None,
         callback_group: CallbackGroup,
         event_callbacks: SubscriptionEventCallbacks,
    ) -> None:
        """
        Create a container for a ROS subscription.

        .. warning:: Users should not create a subscription with this constructor, instead they
           should call :meth:`.Node.create_subscription`.

        :param subscription_impl: :class:`Subscription` wrapping the underlying
            ``rcl_subscription_t`` object.
        :param msg_type: The type of ROS messages the subscription will subscribe to.
        :param topic: The name of the topic the subscription will subscribe to.
        :param callback: A user-defined callback function that is called when a message is
            received by the subscription.
        :param callback_group: The callback group for the subscription. If ``None``, then the
            nodes default callback group is used.
        :param qos_profile: The quality of service profile to apply to the subscription.
        :param raw: If ``True``, then received messages will be stored in raw binary
            representation.
        """
        super().__init__(
            subscription_impl=subscription_impl,
            msg_type=msg_type,
            topic=topic,
            callback=callback,
            qos_profile=qos_profile,
            raw=raw
        )
        self._on_destroy = on_destroy
        self.callback_group = callback_group
        # True when the callback is ready to fire but has not been "taken" by an executor
        self._executor_event = False

        self.event_handlers = event_callbacks.create_event_handlers(
            callback_group, subscription_impl, topic)

    def _destroy(self) -> None:
        for handler in self.event_handlers:
            handler.destroy()
        super()._destroy()

File: rclpy/src/rclpy/subscription.cpp
// Copyright 2021 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <pybind11/pybind11.h>

#include <rcl/error_handling.h>
#include <rcl/node.h>
#include <rcl/subscription.h>
#include <rcl/types.h>
#include <rosidl_runtime_c/message_type_support_struct.h>
#include <rmw/types.h>

#include <memory>
#include <stdexcept>
#include <string>

#include "exceptions.hpp"
#include "node.hpp"
#include "serialization.hpp"
#include "subscription.hpp"
#include "utils.hpp"

using pybind11::literals::operator""_a;

namespace rclpy
{
Subscription::Subscription(
  Node & node, py::object pymsg_type, std::string topic,
  py::object pyqos_profile)
: node_(node)
{
  auto msg_type = static_cast<rosidl_message_type_support_t *>(
    common_get_type_support(pymsg_type));
  if (!msg_type) {
    throw py::error_already_set();
  }

  rcl_subscription_options_t subscription_ops = rcl_subscription_get_default_options();

  if (!pyqos_profile.is_none()) {
    subscription_ops.qos = pyqos_profile.cast<rmw_qos_profile_t>();
  }

  rcl_subscription_ = std::shared_ptr<rcl_subscription_t>(
    new rcl_subscription_t,
    [node](rcl_subscription_t * subscription)
    {
      // Intentionally capture node by copy so shared_ptr can be transferred to copies
      rcl_ret_t ret = rcl_subscription_fini(subscription, node.rcl_ptr());
      if (RCL_RET_OK != ret) {
        // Warning should use line number of the current stack frame
        int stack_level = 1;
        PyErr_WarnFormat(
          PyExc_RuntimeWarning, stack_level, "Failed to fini subscription: %s",
          rcl_get_error_string().str);
        rcl_reset_error();
      }
      delete subscription;
    });

  *rcl_subscription_ = rcl_get_zero_initialized_subscription();

  rcl_ret_t ret = rcl_subscription_init(
    rcl_subscription_.get(), node_.rcl_ptr(), msg_type,
    topic.c_str(), &subscription_ops);
  if (ret != RCL_RET_OK) {
    if (ret == RCL_RET_TOPIC_NAME_INVALID) {
      std::string error_text{"Failed to create subscription due to invalid topic name '"};
      error_text += topic;
      error_text += "'";
      throw py::value_error(error_text);
    }
    throw RCLError("Failed to create subscription");
  }
}

void Subscription::destroy()
{
  rcl_subscription_.reset();
  node_.destroy();
}

py::object
Subscription::take_message(py::object pymsg_type, bool raw)
{
  py::object pytaken_msg;
  rmw_message_info_t message_info;
  if (raw) {
    SerializedMessage taken{rcutils_get_default_allocator()};
    rcl_ret_t ret = rcl_take_serialized_message(
      rcl_subscription_.get(), &taken.rcl_msg, &message_info, NULL);
    if (RCL_RET_OK != ret) {
      if (RCL_RET_BAD_ALLOC == ret) {
        rcl_reset_error();
        throw std::bad_alloc();
      }
      if (RCL_RET_SUBSCRIPTION_TAKE_FAILED == ret) {
        return py::none();
      }
      throw RCLError("failed to take raw message from subscription");
    }
    pytaken_msg = py::bytes(
      reinterpret_cast<const char *>(taken.rcl_msg.buffer),
      taken.rcl_msg.buffer_length);
  } else {
    auto taken_msg = create_from_py(pymsg_type);

    rcl_ret_t ret = rcl_take(
      rcl_subscription_.get(), taken_msg.get(), &message_info, NULL);
    if (RCL_RET_OK != ret) {
      if (RCL_RET_BAD_ALLOC == ret) {
        rcl_reset_error();
        throw std::bad_alloc();
      }
      if (RCL_RET_SUBSCRIPTION_TAKE_FAILED == ret) {
        return py::none();
      }
      throw RCLError("failed to take message from subscription");
    }

    pytaken_msg = convert_to_py(taken_msg.get(), pymsg_type);
  }
  py::object pub_seq_number = py::none();
  if (message_info.publication_sequence_number != RMW_MESSAGE_INFO_SEQUENCE_NUMBER_UNSUPPORTED) {
    pub_seq_number = py::int_(message_info.publication_sequence_number);
  }
  py::object rec_seq_number = py::none();
  if (message_info.reception_sequence_number != RMW_MESSAGE_INFO_SEQUENCE_NUMBER_UNSUPPORTED) {
    rec_seq_number = py::int_(message_info.reception_sequence_number);
  }
  return py::make_tuple(
    pytaken_msg, py::dict(
      "source_timestamp"_a = message_info.source_timestamp,
      "received_timestamp"_a = message_info.received_timestamp,
      "publication_sequence_number"_a = pub_seq_number,
      "reception_sequence_number"_a = rec_seq_number));
}

const char *
Subscription::get_logger_name() const
{
  const char * node_logger_name = rcl_node_get_logger_name(node_.rcl_ptr());
  if (!node_logger_name) {
    throw RCLError("Node logger name not set");
  }

  return node_logger_name;
}

std::string
Subscription::get_topic_name() const
{
  const char * subscription_name = rcl_subscription_get_topic_name(rcl_subscription_.get());
  if (nullptr == subscription_name) {
    throw RCLError("failed to get subscription topic name");
  }

  return std::string(subscription_name);
}

size_t
Subscription::get_publisher_count() const
{
  size_t count = 0;
  rcl_ret_t ret = rcl_subscription_get_publisher_count(rcl_subscription_.get(), &count);
  if (RCL_RET_OK != ret) {
    throw RCLError("failed to get publisher count");
  }

  return count;
}

void
define_subscription(py::object module)
{
  py::class_<Subscription, Destroyable, std::shared_ptr<Subscription>>(module, "Subscription")
  .def(py::init<Node &, py::object, std::string, py::object>())
  .def_property_readonly(
    "pointer", [](const Subscription & subscription) {
      return reinterpret_cast<size_t>(subscription.rcl_ptr());
    },
    "Get the address of the entity as an integer")
  .def(
    "take_message", &Subscription::take_message,
    "Take a message and its metadata from a subscription")
  .def(
    "get_logger_name", &Subscription::get_logger_name,
    "Get the name of the logger associated with the node of the subscription.")
  .def(
    "get_topic_name", &Subscription::get_topic_name,
    "Return the resolved topic name of a subscription.")
  .def(
    "get_publisher_count", &Subscription::get_publisher_count,
    "Count the publishers from a subscription.");
}
}  // namespace rclpy
